packages/lib-core/src/extensions/actions.ts

```diff
@@ -0,0 +1,119 @@
+import type { ReactNode } from 'react';
+import type { ExtensionHook, ExtensionK8sKindVersionModel } from '../api/common-types';
+import type { CodeRef, Extension } from '../types/extension';
+import type { AccessReviewResourceAttributes } from './core-types';
+
+/** ActionProvider contributes a hook that returns list of actions for specific context */
+export type ActionProvider = Extension<
+  'core.action/provider',
+  {
+    /** The context ID helps to narrow the scope of contributed actions to a particular area of the application. Ex - topology, helm */
+    contextId: string | 'resource';
+    /** A react hook which returns actions for the given scope.
+     * If contextId = `resource` then the scope will always be a K8s resource object
+     * */
+    provider: CodeRef<ExtensionHook<Action[]>>;
+  }
+>;
+
+/** ResourceActionProvider contributes a hook that returns list of actions for specific resource model */
+export type ResourceActionProvider = Extension<
+  'core.action/resource-provider',
+  {
+    /** The model for which this provider provides actions for. */
+    model: ExtensionK8sKindVersionModel;
+    /** A react hook which returns actions for the given resource model */
+    provider: CodeRef<ExtensionHook<Action[]>>;
+  }
+>;
+
+/** ActionGroup contributes an action group that can also be a submenu */
+export type ActionGroup = Extension<'core.action/group', Group>;
+
+/** ActionFilter can be used to filter an action */
+export type ActionFilter = Extension<
+  'core.action/filter',
+  {
+    /** The context ID helps to narrow the scope of contributed actions to a particular area of the application. Ex - topology, helm */
+    contextId: string | 'resource';
+    /** A function which will filter actions based on some conditions.
+     * scope: The scope in which actions should be provided for.
+     * Note: hook may be required if we want to remove the ModifyCount action from a deployment with HPA
+     * */
+    filter: CodeRef<(scope: unknown, action: Action) => boolean>;
+  }
+>;
+
+// Type guards
+
+export const isActionProvider = (e: Extension): e is ActionProvider => {
```

Comment:
These type guards are technically boilerplate code, using the following template:

```
export const is${ExtensionTypeName} = (e: Extension): e is ${ExtensionTypeName} => {
  return e.type === ${ExtensionTypeLiteralValue};
};
```

I'd be happier if we could generate them to avoid having to maintain them manually.

@andrewballantyne We should explore using TypeScript custom transformers to generate those type guards during TS compilation. The Rollup plugin we're using `@rollup/plugin-typescript` already supports this TS compiler feature.

https://github.com/madou/typescript-transformer-handbook

Reply:
This can also be handled as part of [HAC-376](https://issues.redhat.com/browse/HAC-376).